core/src/advice/map.rs

```diff
@@ -33,9 +33,31 @@ impl AdviceMap {
         self.0.get(key).map(|v| v.as_slice())
     }
 
-    /// Inserts a key value pair in the advice map and returns the inserted value.
-    pub fn insert(&mut self, key: RpoDigest, value: Vec<Felt>) -> Option<Vec<Felt>> {
-        self.0.insert(key, value)
+    /// Inserts a key-value pair in the advice map.
+    /// Returns None if the insertion was successful, that is if `key` was not present or it was
+    /// already bound to `value`. Returns Some(key, new_value, old_value) in case of failure, that
+    /// is if `key` was already present with a value `old_value` different than `new_value`.
+    pub fn insert(
+        &mut self,
+        key: RpoDigest,
+        value: Vec<Felt>,
+    ) -> Option<(RpoDigest, Vec<Felt>, Vec<Felt>)> {
```

Comment:
nit: I would put a newline between lines 36 and 37 - this will render better in the docs.

Also, the behavior is a bit unusual: typically we'd insert the new value and return the previous value from `insert()` methods. Maybe we should do the same here with the minor modification that if the value was already there we return `None`?

Another alternative could be to have this method return a `Result`.